### PR TITLE
Add text.parse_math rcParams

### PR DESCRIPTION
**Pull request by oscargus**
## PR Summary

Based on the discussion in #22537, there was a question about controlling `parse_math` through rcParams. Assuming that one often would like to use some special packages, it probably would be quite convenient to do that.

Should this be documented as a new feature?

(It is not obvious to me how the default value is controlled...)
 
## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/text.py
Comment (QuLogic) [inline lib/matplotlib/text.py:1295]:
```suggestion
        parse_math : bool, default: :rc:`text.parse_math`
            If False, this `Text` will never use mathtext.  If True, mathtext
            will be used if there is an even number of unescaped dollar signs.
```

File: lib/matplotlib/text.py
Comment (oscargus) [reply lib/matplotlib/text.py:145]: Should I remove the one a few lines up as well?

Comment (QuLogic) [reply lib/matplotlib/text.py:145]: If they're correctly documented.

Comment (oscargus) [reply lib/matplotlib/text.py]: OK! Two additional comment removed and `set_color` slightly better documented.

**Comment by oscargus** [comment]
Just to confirm, in general the preferred approach is as. 

In `__init__(..., parameter=None, ...)`:
```
self._parameter = None 
self.set_parameter(parameter)
```

In `set_parameter`:
```
...
     parameter : type, default: :rc:`...`
...
if parameter is None:
    self._parameter = rcParams...
else:
    self._parameter = parameter
```
?

I noted that this didn't really hold in all files. Main benefit, I guess, is that the default value is handled consistently both in documentation and in the code (so that it is "always" possible to call with None to reset to default) and so that the redundant calls to rcParam are reduced.

**Comment by oscargus** [comment]
Modified the test to use rc_context (which I assume is the better way?).

**Comment by timhoffm** [comment]
> Modified the test to use rc_context (which I assume is the better way?).

Each test gets fresh rcParams to ensure they are isolated. So it doesn't really matter here.

> Just to confirm, in general the preferred approach is as.

We do not have a policy on this. This approach makes the setters to accept `None`, meaning "set to rcParams". This is a bit of a stretch of the setter concept.

File: lib/matplotlib/text.py
Comment (timhoffm) [inline lib/matplotlib/text.py:1274]: This line is both used in the [docstring of the setter](https://67257-1385122-gh.circle-artifacts.com/0/doc/build/html/api/text_api.html?highlight=text#matplotlib.text.Text.set_usetex) and in the listing of the properties (e.g. [here](https://67257-1385122-gh.circle-artifacts.com/0/doc/build/html/api/text_api.html?highlight=text#matplotlib.text.Text).

While `bool, default: ...` makes sense in the latter, it does for the setter. The setter has no default. You cannot do `set_usetex()`. To accommodate both cases, we should stick to the original `bool or None` and describe what None means verbally. That's not as nice for the listing, but at least it's correct. I'm afraid we'd need some more machinery if we wanted to be be better here.

Same holds for the other doc changes in this PR.

**Comment by oscargus** [comment]
> This is a bit of a stretch of the setter concept.

In a good or a bad way? 😄 

Just noted that this was heavily used in the rest of this file (at least). 

(Saw on SE (although I cannot find it now) that the `self._parameter = None ` in `__init__` was not really encouraged. I seem to recall that some static code analysis tools are complaining, but maybe they realize that it in fact is initialized when `__init__` is called.)

File: lib/matplotlib/text.py
Comment (oscargus) [reply lib/matplotlib/text.py:1274]: I see your point. However, as the kwdoc is use in the listing of the properties, this seems to be the only way to get them there. (One way is of course to change it to a keyword-argument with default None, in that way it is slightly more correct...).

Also, see https://github.com/matplotlib/matplotlib/pull/22556#r816451833 (which I guess is why I did it).

Comment (timhoffm) [reply lib/matplotlib/text.py:1274]: What I'm trying to say is that there is no good solution here. Setters and passing by keyword are conceptually different enough so that there is no wording that can correctly describe both in detail.

### Option 1 (original)

```
def set_usetex(self, usetex)
    """
    ...
    usetex : bool or None
        Whether to render using TeX, ``None`` means to use :rc:`text.usetex`.
```
**Advantage:** The setter has a reasonable API and is documented correctly. `set_usetex(None)` meaning use the default `rcParams` is a bit of a stretch but bearable (and we have precedence so it's not that easy to change).
**Disadvantage:** Property lists only show "bool or None", you have to click through to the setter to find the rcParams name and value.

### Option 2 (current PR)

```
def set_usetex(self, usetex):
    """
    ...
    usetex : bool, default: :rc:`text.usetex`
        Whether to render using TeX.
```
**Advantage:** Property lists show nicely as "bool or rcParams['text.usetex'], default: ..."
**Disadvantage:** The setter description is incorrect. usetex has no default, instead one has to pass None, which is not documented here.

### Option 3 (the above comment)

```
def set_usetex(self, usetex=None):
    """
    ...
    usetex : bool, default: :rc:`text.usetex`
        Whether to render using TeX.
```
**Advantage:** Property lists show nicely as "bool or rcParams['text.usetex'], default: ...". The setter description is correct.
**Disadvantage:** The setter API is awkward. Having `set_usetex()` mean "use the rcParams default" is quite opaque.

## Summary
Of these option 1 is the least bad. If you want to have better property listing, that must use something else than only taking the setter type description.

Comment (oscargus) [reply lib/matplotlib/text.py:1274]: Thanks for the well-made outline of the options! I do indeed see the problem. Maybe @QuLogic would like to chip in here as I interpreted his suggestion in https://github.com/matplotlib/matplotlib/pull/22556#discussion_r821544848 as he preferred Option 2.

We should probably discuss this in a dev call as it affects some other PRs as well (not to mention upcoming ones).

Comment (QuLogic) [reply lib/matplotlib/text.py:1274]: If this needs some special casing, we need to update [the documentation](https://matplotlib.org/devdocs/devel/documenting_mpl.html#default-values).

Comment (timhoffm) [reply lib/matplotlib/text.py:1274]: The other question would be whether such setters should accept None as "use rcparams" at all or whether defaults should be handled separately in __init__ (explicitly or using to be defined helpers).

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
I think we should have another reviewer because the PR changed significantly after @tacaswell's approval.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**
